src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -1157,38 +1157,46 @@ public Object initializeCopy(DynamicObject self, DynamicObject from) {
     @ImportStatic(StringGuards.class)
     public abstract static class LstripBangNode extends CoreMethodArrayArgumentsNode {
 
+        @Child private RopeNodes.GetCodePointNode getCodePointNode = RopeNodes.GetCodePointNode.create();
         @Child private RopeNodes.MakeSubstringNode makeSubstringNode = RopeNodes.MakeSubstringNode.create();
 
         @Specialization(guards = "isEmpty(string)")
         public DynamicObject lstripBangEmptyString(DynamicObject string) {
             return nil();
         }
 
-        @TruffleBoundary
         @Specialization(guards = { "!isEmpty(string)", "isSingleByteOptimizable(string)" })
-        public Object lstripBangSingleByte(DynamicObject string) {
+        public Object lstripBangSingleByte(DynamicObject string,
+                                           @Cached("create()") RopeNodes.BytesNode bytesNode,
+                                           @Cached("createBinaryProfile()") ConditionProfile noopProfile) {
             // Taken from org.jruby.RubyString#lstrip_bang19 and org.jruby.RubyString#singleByteLStrip.
 
             final Rope rope = rope(string);
-            final int s = 0;
-            final int end = s + rope.byteLength();
-            final byte[] bytes = rope.getBytes();
+            final int firstCodePoint = getCodePointNode.executeGetCodePoint(rope, 0);
 
-            int p = s;
-            while (p < end && ASCIIEncoding.INSTANCE.isSpace(bytes[p] & 0xff)) p++;
-            if (p > s) {
-                StringOperations.setRope(string, makeSubstringNode.executeMake(rope, p - s, end - p));
+            // Check the first code point to see if it's a space. In the case of strings without leading spaces,
+            // this check can avoid having to materialize the entire byte[] (a potentially expensive operation
+            // for ropes) and can avoid having to compile the while loop.
+            if (noopProfile.profile(!StringSupport.isAsciiSpace((byte) firstCodePoint))) {
+                return nil();
+            }
 
-                return string;
+            final int end = rope.byteLength();
+            final byte[] bytes = bytesNode.execute(rope);
+
+            int p = 0;
```

Comment:
Minor: you could start at 1 here since you already checked the first byte.

Reply:
Indeed. Although I thought HotSpot was able to compile loops that start at 0 a bit better. Maybe that's not accurate or not really our concern.

Reply:
In this case what matters is how Truffle compiles it and I don't think there is any difference.